Support EXT_disjoint_timer_query_webgl2

mosra opened the pull request:
Continuation of #4575 (from 2016), I just ran into this today when adding support for this extension into [Magnum](https://magnum.graphics).

Until now, most of the extension worked when using the core query APIs, except for `glQueryCounterEXT()` and `glGetQueryObjecti64vEXT()` -- there the functions assumed `glGenQueriesEXT()` was called that populated the internal id-to-object mapping array (and not `glGenQueries()`). So I added branches there that do the correct thing under WebGL 2.

Moreover, the `queryCounterEXT` property was undefined because the code asked for `EXT_disjoint_time_query` on WebGL 2 (which isn't exposed there). The code now asks for `EXT_disjoint_timer_query_webgl2` there, which makes `queryCounterEXT` available.

Finally, because code coming from GLES may expect the EXT-suffixed versions still being available on WebGL 2 builds as well (since you can use the EXT-suffixed `glGenQueriesEXT()` etc. with `EXT_disjoint_time_query` on ES3 well), those are now aliases to the core APIs.

I did quite an extensive testing with Magnum's GL tests (both WebGL 1 and WebGL 2), but it may happen that this change will affect some Emscripten tests. Also, if there's any test file that I could update to cover the now-working APIs for WebGL 2, let me know.

EDIT: upon discovering that Firefox doesn't actually implement the `_webgl2` version on WebGL 2 (https://bugzilla.mozilla.org/show_bug.cgi?id=1328882), I added a fallback to the WebGL 1 extension string. As far as my testing goes, the entry points work correctly, so it's just about the extension string.

Thank you!

kainino0x commented on src/library_webgl.js:
```diff
@@ -1784,35 +1798,50 @@ var LibraryGL = {
     GL.validateGLObjectID(GL.timerQueriesEXT, id, 'glQueryCounterEXT', 'id');
 #endif
+#if USE_WEBGL2
+    GLctx.disjointTimerQueryExt['queryCounterEXT'](GL.queries[id], target);
```
Yes. Both extensions expose the same queryCounterEXT function.

juj replied:
I think instead of using two arrays `GL.queries` and `GL.timerQueriesEXT`, let's just use one array to simplify the code, i.e. let's remove `GL.timerQueriesEXT` array altogether, and lump that in with `GL.queries`? I don't think there is a scenario where we should distinguish between these two.

mosra replied:
Done -- `GL.timerQueriesEXT` renamed to `GL.queries`.

kainino0x commented on src/library_webgl2.js:
```diff
@@ -244,13 +247,15 @@ var LibraryWebGL2 = {
   glIsQuery: function(id) {
     var query = GL.queries[id];
     if (!query) return 0;
     return GLctx['isQuery'](query);
```
For USE_WEBGL2 + webgl1, this and others in this file wouldn't be present.

kainino0x replied:
Actually, this file probably wouldn't even be included. So you'll need to instead move the definitions into library_webgl.js, and those definitions can choose at runtime between webgl1+ext and webgl2.

mosra replied:
I'm no longer doing any EXT aliases on WebGL 2 anymore, so this is resolved as well.

kainino0x commented on src/library_webgl.js:
```diff
@@ -1867,8 +1902,13 @@ var LibraryGL = {
 #endif
+#if USE_WEBGL2
+    var query = GL.queries[id];
+    var param = GLctx['getQueryParameter'](query, pname);
```
For USE_WEBGL2 + webgl1, getQueryParameter wouldn't be present here.

mosra replied:
Done -- this is a runtime check now.

mosra commented on src/library_webgl.js:
```diff
@@ -960,6 +960,14 @@ var LibraryGL = {
+      // on core query APIs. HOWEVER, Firefox implements only the WebGL 1
+      // version and thus we look for the WebGL 2 version only if the WebGL 1
+      // version isn't present. See also https://bugzilla.mozilla.org/show_bug.cgi?id=1328882
+      if (!GLctx.disjointTimerQueryExt)
```
Will do.

mosra commented:
@kainino0x Apologies -- I wasn't aware creating a WebGL 1 context with `USE_WEBGL2=1` was even possible :) Any ideas on how should I proceed? If I see it correctly:

- for WebGL 1 builds `#ifdef`ing the WebGL2-specific parts
- for WebGL 2 builds, it would all need to decide at runtime, based on which context is created, right? What's the "canonical" way to do it?

Thank you!

kainino0x commented:
> I wasn't aware creating a WebGL 1 context with USE_WEBGL2=1 was even possible

I'm not 100% sure it is, but I think it is. Hoping @juj would chime in.

I can look into it though.

> EDIT: Oh and re tests -- there's a problem that Firefox exposes the extension only under a flag (not sure how you can enable that on the CIs), and

(FYI, I believe this is a Spectre mitigation.)
I don't know anything about Emscripten's CI so not sure if we'd be able to test with a flag, but @kripken certainly knows.

> both chrome and firefox report 0s unless the results are queried async. It could check for GL errors at least, tho.

Yup, that's intended behavior. A test can just use emscripten_set_main_loop or similar to allow the browser to update the value.

kainino0x commented:
It's also very valuable to have tests even if we have to run them manually/locally, so it would be great to have them.

kripken commented:
Testing with a firefox pref is possible, see here where we already set up a few prefs: https://github.com/emscripten-core/emscripten/blob/ea3209ae0427a8223e7ff108f84b55d05986b8e1/.circleci/config.yml#L131

kainino0x commented:
> I can look into it though.

Tried a basic emscripten_webgl_init_context_attributes with USE_WEBGL2=1 and attrs.majorVersion = 1, and that works on Safari (which doesn't have WebGL 2). Didn't try anything more complex.

kainino0x commented:
@mosra 
> * for WebGL 2 builds, it would all need to decide at runtime, based on which context is created, right? What's the "canonical" way to do it?

I think simply with conditionals like `if (GL.currentContext.version >= 2)`.

mosra commented:
Sorry for the silence on my end -- had two extremely busy weeks. Hoping to get back to this over the weekend or early next week.

juj commented on src/library_webgl2.js:
```diff
@@ -298,6 +305,8 @@ var LibraryWebGL2 = {
     {{{ makeSetValue('params', '0', 'ret', 'i32') }}};
   },
+  glGetQueryObjectivEXT: 'glGetQueryObjectuiv',
+  glGetQueryObjectuivEXT: 'glGetQueryObjectuiv',
```
Each aliased function should have a corresponding `funcname__sig: 'signaturehere'` field that specifies the function signatures of the aliases. That will help the aliasing backend use a better implemented aliasing function.

mosra replied:
Done (the aliases are no longer there, I chose a different approach).

juj commented on src/library_webgl.js:
```diff
@@ -1784,35 +1798,50 @@ var LibraryGL = {
   glEndQueryEXT: function(target) {
     GLctx.disjointTimerQueryExt['endQueryEXT'](target);
   },
+#endif
```
Generally we have reused the `#if !USE_WEBGL2` blocks across multiple functions, so can just have this once for all of these.

mosra replied:
Done -- actually I removed most of the preprocessor blocks now.

sbc100 commented:
The deletion of the `incoming` branch triggered the automatic closing of this branch.  My apologies.  This was an unexpected side effect.

Re-opening and re-targeting to master.

mosra commented:
Sorry for the over-a-year-long delay, finally found the time to wrap this up (also a bit frightened that nobody else needed this functionality yet, huh). 

I rethought the original idea a bit and the actual change in functionality is a lot smaller than before:

- there's now tests for timer queries in WebGL 1, WebGL 2 and a WebGL 2 build with a WebGL 1 context
  - and sample queries in WebGL 2, because I realized that one way to fix time queries on WebGL 2 would break the other queries there, so better to have both covered
- removed the EXT aliases in WebGL 2 as those are not essential for my use case (and considering nobody else expressed interest in this during the year this PR was open, I doubt anybody else needs them either); plus it would make the implementation unnecessarily complex
- as @juj suggested, I removed the `GL.timerQueriesEXT` in favor of `GL.queries` that's now used in both WebGL 1 and WebGL 2
- on WebGL 2 builds, the `glGetQueryObjectui64vEXT()` does a runtime check and uses appropriate functionality depending on whether WebGL 1 or WebGL 2 context is created
- on WebGL 2, `EXT_disjoint_timer_query_webgl2` is requested instead of `EXT_disjoint_timer_query`, together with  a fallback for Firefox which exposes only `EXT_disjoint_timer_query` on both

What isn't in the tests here is the `glQueryCounterEXT()` function, as I had a difficulty with it (returning 0 on Chrome on WebGL 1, causing a GL error on Chrome on WebGL 2; crashing the tab on Firefox). Not sure what's up, will investigate further and open a followup PR with tests covering this API (and exposing this problem) once this one is merged.

kainino0x commented on tests/test_browser.py:
```diff
@@ -4265,6 +4265,24 @@ def test_webgl_draw_base_vertex_base_instance(self):
+  def test_webgl_timer_query(self):
+    for args in [
+        # EXT query entrypoints on WebGL 1.0
+        ['-s', 'MAX_WEBGL_VERSION'],
```
I'm not sure, is this the same as `MAX_WEBGL_VERSION=1`? Could we just explicitly specify `=1`?

mosra replied:
Honestly I have no idea. Copypasted it from another occurence in this file which led me to believe it indeed is an implicit `=1` :)

kripken replied:
Yes, emcc allows `-s X` instead of `-s X=1`.

kainino0x reviewed (commented):
Haven't had a chance to read through the tests yet, but I think this looks good! Great work with the thorough testing, based on your description - that really helps be confident things work and won't break later.

kainino0x commented on tests/webgl_timer_query.cpp:
```diff
@@ -0,0 +1,165 @@
+#ifdef TEST_WEBGL2
+  GL_CALL(glDeleteQueries(1, &timerQuery));
+#else
+  GL_CALL(glDeleteQueriesEXT(1, &timerQuery));
```
Should either of these work on the other context - i.e. glDeleteQueries when using WebGL 1 or glDeleteQueriesEXT when using WebGL2? I think at least the latter should work, if not both.

Would be nice to test this. We could do this by splitting TEST_CORE_QUERIES from TEST_WEBGL2  and testing 3-4 combinations of them. WDYT?

mosra replied:
Haha, sorry, I already swapped out all knowledge I had about this... but if my sparse memory serves me right, it's not interchangeable:

- the EXT version works only on WebGL 1 contexts as it comes from the [EXT_disjoint_timer_query](https://www.khronos.org/registry/webgl/extensions/EXT_disjoint_timer_query/) (which, according to the spec, shouldn't even be exposed on WebGL 2)
- the unsuffixed version works only on WebGL 2 contexts as [EXT_disjoint_timer_query_webgl2](https://www.khronos.org/registry/webgl/extensions/EXT_disjoint_timer_query_webgl2/) relies on the core WebGL 2 query functionality for everything except 64-bit query result retrieval

Maybe you're referring to the original version of this PR (from 2019) which had some compatibility bits to make the EXT variants working on WebGL 2 as well (and thus mirroring the ES [EXT_disjoint_timer_query](https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_disjoint_timer_query.txt), which allows the EXT functions on ES3)? I dropped that since, because the maintenance overhead didn't seem to be worth it (and I didn't have any need for it in Magnum either).

kainino0x replied:
Got it, I think you're right I got confused with the original version. Just wanted to make sure we're testing whatever is implemented.

kainino0x reviewed (approved):
Sorry for the delay on this. Tests LGTM, though I had an idea about how we could make them a little more thorough. Thanks again for working on this!

Let us know if you won't have time to do this and I think I could pick up the change I suggested.

mosra commented:
Thank you!
